### ISSUE TITLE
Detect variables with the same name as functions and  throw an exception at "compile" time

### ISSUE THREAD

**Issue by teo1978** [processing-js/processing-js#181]
Similar but not the same as another issue that I have recently reported: it is known and documented that there are caveats when porting Processing code to PJS. However, the lack of any attempt whatsoever to detect them makes them a hell to debug.

This time, I had a class with a property called 'randomSeed'. Then I also called randomSeed() (a Processing core function) from a method of the same class. The result was that code which runs in Java mode would throw an impossible-to-debug exception:
  Uncaught TypeError: $this_1.randomSeed is not a function

At "compile" time, Processing.js should be able to detect when variables are defined with the same name as existing functions, and throw an exception before running the sketch, which could be much easier to debug, such as "You cannot define a variable named *xxxxx* because a function of the same name already exist".

I guess it may not be trivial, but there are at least two cases that should not be that difficult to detect and that may cover a great percentage of why-doesnt-this-work-in-pjs use cases:
1) properties and methods with the same name in the same class
2) variables with the same name as "global" processing functions (i.e. methods of the PApplet class)

**Comment by Pomax** [comment]
Note that instances of (2) only interfere if they declared at the same level, so this will break:

```
int size = 10; // overwrites the globally scoped "size()" function
void setup() {
  size(100,100); // number is not  a function
}
```
because you're creating a global variable that conflicts a preexisting global. However, this works fine:
```
void setup() {
  int println = 10; // scoped, the original println is merely shadowed inside this function
}
void draw() {
  println("testing");
}
```
Although this will be a problem because of JS's "everything is a property" rather than distinguishing between variables and functions:
```
void setup() {
  int size = 400;
  size(100,100); // no separate symbol tables for functions vs. not functions: number is not a function
}
```

**Comment by Pomax** [comment]
To add to why these things are actually fairly difficult: Processing.js does the minimum amount of rewriting to turn Processing syntax into JavaScript syntax. This includes blind transforms where what to human eyes looks like an obvious thing (like "int somename = 3") is never even inspected by the parser (getting blind-rewritten to "var somename = 3", for instance, without ever seeing the word "somename" as a distinct string). That makeas it very hard to check whether variable or function names collide with the API's naming, although there might be some parts in [the Parser object](https://github.com/processing-js/processing-js/blob/master/src/Parser/Parser.js) where we can try to capture variable and method names and do a sanity check, but the code was written by someone much more capable at writing transpilers than I am, and anything but easy to understand from the times I had to work on it.

**Comment by teo1978** [comment]
> Note that instances of (2) only interfere if they declared at the same level

Not exactly. See the example I cited:
> This time, I had a class with a property called 'randomSeed'. Then I also called randomSeed() (a Processing core function) from a method of the same class. The result was that code which runs in Java mode would throw an impossible-to-debug exception:
Uncaught TypeError: $this_1.randomSeed is not a function

Just in case it was unclear (this is not my actual code, just a simplified example):
```
class TestClass {
  int randomSeed
  
  TestClass(String a) {
    randomSeed=a.length();
  }
  
  public void makeMeCrash() {
    randomSeed(randomSeed); // produces error
  }
}
```
I do understand why it happens, I'm just saying it's not "declared at the same level".

**Comment by teo1978** [comment]
> Processing.js does the minimum amount of rewriting to turn Processing syntax into JavaScript syntax. This includes blind transforms where what to human eyes looks like an obvious thing is never even inspected by the parser

Yeah, I pretty much guessed that. I think it's necessary for the parser to inspect a little bit more and be a little less blind.

**Comment by Pomax** [comment]
Parsers are ludicrously hard things, but it's worth seeing if we can even find the place where it might know about variable and function names.
